internal/alb/sg/instance_attachment.go

```diff
@@ -0,0 +1,193 @@
+func (controller *instanceAttachmentController) findENIsSupportingTargets(instanceENIs map[string][]*ec2.InstanceNetworkInterface, targets tg.TargetGroups) map[string]bool {
+	result := make(map[string]bool)
+	for _, group := range targets {
+		if group.TargetType == elbv2.TargetTypeEnumInstance {
+			for _, eniID := range controller.findENIsSupportingTargetGroupOfTypeInstance(instanceENIs, group) {
+				result[eniID] = true
+			}
+		} else {
+			for _, eniID := range controller.findENIsSupportingTargetGroupOfTypeIP(instanceENIs, group) {
+				result[eniID] = true
+			}
+		}
+	}
+	return result
+}
+
+// findENIsSupportingTargetGroupOfTypeInstance find the ID of ENIs that are used to supporting ingress traffic to targetGroup with targetType instance.
+// For targetType instance, traffic is routed into primary ENI of instance(eth0, i.e. decviceIndex == 0), other network interfaces are not used.
+func (controller *instanceAttachmentController) findENIsSupportingTargetGroupOfTypeInstance(instanceENIs map[string][]*ec2.InstanceNetworkInterface, group *tg.TargetGroup) (result []string) {
+	targetInstanceIDs := make(map[string]bool)
+	for _, endpoint := range group.DesiredTargets() {
+		targetInstanceIDs[aws.StringValue(endpoint.Id)] = true
+	}
+	for instanceID, enis := range instanceENIs {
+		if _, ok := targetInstanceIDs[instanceID]; !ok {
+			continue
+		}
+		for _, eni := range enis {
+			if aws.Int64Value(eni.Attachment.DeviceIndex) == 0 {
+				result = append(result, aws.StringValue(eni.NetworkInterfaceId))
+			}
+		}
+	}
+	return result
+}
+
+// findENIsSupportingTargetGroupOfTypeIP find the ID of ENIs that are used to supporting ingress traffic to targetGroup with targetType IP.
+// For targetType IP, traffic is routed into the ENI for specific pod IPs.
+// Warning: this function only works under CNI implementations that use ENI for pod IPs such as amazon k8s cni.
+func (controller *instanceAttachmentController) findENIsSupportingTargetGroupOfTypeIP(instanceENIs map[string][]*ec2.InstanceNetworkInterface, group *tg.TargetGroup) (result []string) {
+	targetPodIPs := make(map[string]bool)
+	for _, endpoint := range group.DesiredTargets() {
+		targetPodIPs[aws.StringValue(endpoint.Id)] = true
+	}
+	for _, enis := range instanceENIs {
+		for _, eni := range enis {
+			for _, addr := range eni.PrivateIpAddresses {
+				if _, ok := targetPodIPs[aws.StringValue(addr.PrivateIpAddress)]; ok {
+					result = append(result, aws.StringValue(eni.NetworkInterfaceId))
+					break
+				}
+			}
+		}
+	}
+	return result
+}
+
+// getClusterInstanceENIs retrives all ENIs attached to instances indexed by instanceID
+func (controller *instanceAttachmentController) getClusterInstanceENIs() (map[string][]*ec2.InstanceNetworkInterface, error) {
+	instanceIDs, err := controller.getClusterInstanceIDs()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get instance IDs within cluster, Error:%s", err.Error())
```

Comment:
nits: `s/Error:%s/Error: %s/`?

Reply:
actually it probably doesn't even need the `Error:`

Reply:
i have already changed these to be "XXXX due to %s" 😄